internals/overlord/logstate/loki/loki_test.go

```diff
@@ -0,0 +1,204 @@
+  "stream": {"pebble_service": "svc2"},
+  "values": [
+      [ "1704026091000000000", "log line #2" ],
+      [ "1704026096000000000", "log line #7" ]
+  ]
+}, {
+  "stream": {"pebble_service": "svc3"},
+  "values": [
+      [ "1704026093000000000", "log line #4" ],
+      [ "1704026095000000000", "log line #6" ]
+  ]
+}, {
+  "stream": {"pebble_service": "svc4"},
+  "values": [
+      [ "1704026098000000000", "log line #9" ]
+  ]
+}]}`
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		c.Assert(r.Method, Equals, http.MethodPost)
+		c.Assert(r.Header.Get("Content-Type"), Equals, "application/json; charset=utf-8")
+
+		reqBody, err := io.ReadAll(r.Body)
+		c.Assert(err, IsNil)
+		expFlattened, err := flattenJSON(expected)
+		c.Assert(err, IsNil)
+		c.Assert(string(reqBody), Equals, expFlattened)
+	}))
+	defer server.Close()
+
+	client := loki.NewClient(&plan.LogTarget{Location: server.URL})
+	for _, entry := range input {
+		err := client.Add(entry)
+		c.Assert(err, IsNil)
+	}
+
+	err := client.Flush(context.Background())
+	c.Assert(err, IsNil)
+}
+
+func (*suite) TestFlushCancelContext(c *C) {
+	serverCtx, killServer := context.WithCancel(context.Background())
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		select {
+		case <-serverCtx.Done():
+		// Simulate a slow-responding server
+		case <-time.After(10 * time.Second):
+		}
+	}))
+	defer server.Close()
+	defer killServer()
+
+	client := loki.NewClient(&plan.LogTarget{Location: server.URL})
+	err := client.Add(servicelog.Entry{
+		Time:    time.Now(),
+		Service: "svc1",
+		Message: "this is a log line\n",
+	})
+	c.Assert(err, IsNil)
+
+	flushReturned := make(chan struct{})
+	go func() {
+		// Cancel the Flush context quickly
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Microsecond)
+		defer cancel()
+
+		err := client.Flush(ctx)
+		c.Assert(err, ErrorMatches, ".*context deadline exceeded.*")
+		close(flushReturned)
+	}()
+
+	// Check Flush returns quickly after context timeout
+	select {
+	case <-flushReturned:
+	case <-time.After(1 * time.Second):
+		c.Fatal("lokiClient.Flush took too long to return after context timeout")
+	}
+}
+
+func (*suite) TestServerTimeout(c *C) {
+	stopRequest := make(chan struct{})
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		<-stopRequest
+	}))
+	defer server.Close()
+	defer close(stopRequest)
+
+	client := loki.NewClientWithArgs(
+		&plan.LogTarget{Location: server.URL},
+		loki.ClientArgs{
+			RequestTimeout: 1 * time.Microsecond,
+		},
+	)
+	err := client.Add(servicelog.Entry{
+		Time:    time.Now(),
+		Service: "svc1",
+		Message: "this is a log line\n",
+	})
+	c.Assert(err, IsNil)
+
+	err = client.Flush(context.Background())
+	c.Assert(err, ErrorMatches, ".*context deadline exceeded.*")
+}
+
+// Strips all extraneous whitespace from JSON
+func flattenJSON(s string) (string, error) {
```

Comment:
Is this basically [`json.Compact`](https://pkg.go.dev/encoding/json#Compact)? Could just do this inline:

```go
var buf bytes.Buffer
err = json.Compact(&buf, []byte(s))
// use buf.String()
```

Reply:
Nice, wasn't aware of that one.